### ISSUE TITLE
Mosaic Augmentation for Object Detection 

### ISSUE THREAD

**Issue by innat** [keras-team/keras-cv#21]
**Mosaic Augmentation**
- Paper: [YOLOv4: Optimal Speed and Accuracy of Object Detection](https://arxiv.org/pdf/2004.10934.pdf) (Figure 3).
- **Yolo-V4** Citation: 3718

**What is it?** 
The idea is to take 4 random samples and create a single sample of mosaic fashion with them. For example:

![mosaic](https://user-images.githubusercontent.com/17668390/145579025-cb51323d-d582-4984-895a-0844a6d34d52.png)

**Starter** 

- PyTorch Implementation: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1YWb7a_3bHqG30SoIxU5S4lHKktkRseyY?usp=sharing)
- TensorFlow Implementation: [google/automl/efficientdet/aug/mosaic.py](https://github.com/google/automl/blob/master/efficientdet/aug/mosaic.py)

**Comment by bhack** [comment]
I think we have already in the official model garden:
https://github.com/tensorflow/models/blob/master/official/vision/beta/projects/yolo/ops/preprocessing_ops.py#L236
https://github.com/tensorflow/models/blob/master/official/vision/beta/projects/yolo/ops/mosaic.py

**Comment by innat** [comment]
Oh, I see. Does it mean that we should use these from the `tensorflow/models` garden or all those relevant operations will be moved here to strengthen the high-level api? Also why the `mosaic.py` file is under `/yolo/ops/` directory? Shouldn't it be more general for all object detection models?

**Comment by bhack** [comment]
This is the old long story about where to collect reusable CV components. 
I suppose that we could port these from the model garden to keras-cv.

**Comment by LukeWood** [comment]
Think this looks like another good preprocessing layer.

> I suppose that we could port these from the model garden to keras-cv.
yeah I agree, it's beginning to look like we are probably going to have a lot of code duplication between keras-cv and model garden.  In my opinion this is to be expected - We do not want keras-cv to depend on model garden, but some of the functionality is the same.

@bhack thank you for continuing to point out  the existing model garden implementations.  They're extremely helpful to me.

**Comment by innat** [comment]
@bhack @LukeWood @qlzh727 
What should be the **default target format** for object detection tasks? Will it be pascal format or coco format? In the object detection preprocessing layer, is it expected that the layer should support both formats as input?

**Comment by innat** [comment]
@LukeWood could you please see the above query? I would most likely send a PR for mosaic if possible.

**Comment by LukeWood** [comment]
Good question, probably corners?  This can have its own issue to discuss.  I'll re-open!
